### PR TITLE
Metrics Server: use gRPC connection to get metrics from Operator

### PR DESCRIPTION
**Pull request by zroubalik**
Signed-off-by: Zbynek Roubalik <zroubalik@gmail.com>

**Changes**
- gRPC Server runs on Operator on port 9666 and provides metrics for Metrics Server, we are reusing scalers cache (with a connection to external service) that is already present on Operator, this should reduce the number of opened connections from KEDA approximately by half.
- The old approach (with querying metrics directly from Metrics Server) is maintained and could be enabled on Metrics Server by setting env variable `KEDA_USE_METRICS_SERVICE_GRPC` to `false`, this option is deprecated and should be removed in the future releases
- Prometheus Metrics on Operator now provides all metrics, Metrics Server still provides the metrics it did before, this option is deprecated and should be removed in the future releases
- remove `MetricSelector` field from `Scaler` interface (and thus from all scalers) since it hasn't had really any use
- moved `fallback` out of the `provider` package

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [x] A PR is opened to update our Helm chart https://github.com/kedacore/charts/pull/337
- [x] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs/pull/989
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes https://github.com/kedacore/keda/issues/3920
Fixes https://github.com/kedacore/keda/issues/3919

Relates to https://github.com/kedacore/keda/issues/2282

### CODE REVIEW

File: pkg/scaling/scale_handler.go
Comment (JorTurFer) [inline pkg/scaling/scale_handler.go:184]: why is this different from the if above? I mean, this condition will be always false because if it's true, the flow will return with the if above

File: pkg/scaling/scale_handler.go
Comment (JorTurFer) [inline pkg/scaling/scale_handler.go:188]: why is this different from the if above? I mean, this condition will be always false because if it's true, the flow will return with the if above.
Apart from that, both branches test the same `ok`, so logically this won't happen never I'd say

Comment (zroubalik) [reply pkg/scaling/scale_handler.go:188]: good catch, this is just a bad copy paste

**Comment by zroubalik** [comment]
> /run-e2e **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/3483698301)

🎉

The only failed test is `./internals/prometheus_metrics/prometheus_metrics_test.go` which is expected, because the Prometheus stuff is not yet migrated.

**Comment by zroubalik** [comment]
We are no longer querying the metrics values in Metrics Server but in the Operator and this affects exposed Prometheus metrics https://keda.sh/docs/2.8/operate/prometheus/#metrics-adapter

I'd deprecate the exposed metrics in Metrics Server and migrate them to Operator -> this way we have one place for all Prom metrics - the Operator https://github.com/kedacore/keda/issues/1281

I'd like to avoid transferring the metrics locally from Operator to MS, it could be done to maintain the current state, but I don't like this solution. 

@kedacore/keda-maintainers WDYT?

**Comment by JorTurFer** [comment]
> I'd like to avoid transferring the metrics locally from Operator to MS, it could be done to maintain the current state, but I don't like this solution.
> 
> @kedacore/keda-maintainers WDYT?

I'd document this as a breaking change and I'd continue, this is a really nice improvement for the performance and if we maintain the same names and we update the annotations in deployment, metrics will be there...

**Comment by zroubalik** [comment]
> I'd document this as a breaking change and I'd continue, this is a really nice improvement for the performance and if we maintain the same names and we update the annotations in deployment, metrics will be there...

Yeah, another big plus for me is that we will have all Prom metrics in one place. And if we implement the multi-instances support for cluster, it would benefit from this also. We can always introduce some other Prom metrics on MS side if needed.

**Comment by zroubalik** [comment]
I will also send PR for docs and chart. Once this is merged I will try to tackle the caching part.

**Comment by zroubalik** [comment]
> I'm going to tackle [this](https://github.com/kedacore/keda/issues/3927) e2e improvement, hopefully today, does it make sense to wait till is done, in order to test this change with that test?

I'd prefer if we go ahead and merge this, I test that manually and it is working for me. I would like to follow up with other features and don't want to make this PR huge (as it already is :)) )

**Comment by JorTurFer** [comment]
okey, I'll review this PR before starting with the other issue later on today

**Comment by JorTurFer** [comment]
Should we update docs to explain `KEDA_USE_METRICS_SERVICE_GRPC`?

File: pkg/prommetrics/prommetrics.go
Comment (JorTurFer) [inline pkg/prommetrics/prommetrics.go:115]: Can we remove this commented code?

**Comment by zroubalik** [comment]
> Should we update docs to explain `KEDA_USE_METRICS_SERVICE_GRPC`?

Yeah, that's the plan, I will send docs PR later today.

File: pkg/scaling/scale_handler.go
Comment (JorTurFer) [inline pkg/scaling/scale_handler.go:55]: Can we remove this comment?

**Review by JorTurFer — APPROVED** [review]
Awesome job!!!
only some nits inline, apart from them, could we add the version when the code will be removed or when has been deprecated in the `[[ Deprecated ]]` comments? just to remember it, something like `[[ Deprecated in v2.9]]`

**Comment by zroubalik** [comment]
> Awesome job!!! only some nits inline, apart from them, could we add the version when the code will be removed or when has been deprecated in the `[[ Deprecated ]]` comments? just to remember it, something like `[[ Deprecated in v2.9]]`

I will create an issue to track this.

**Comment by zroubalik** [comment]
> Awesome job!!! only some nits inline, apart from them, could we add the version when the code will be removed or when has been deprecated in the `[[ Deprecated ]]` comments? just to remember it, something like `[[ Deprecated in v2.9]]`

https://github.com/kedacore/keda/issues/3930

**Comment by zroubalik** [comment]
@v-shenoy would be great if you can take a look on this, more eyes more see. We would like to merge this soon and follow up.

**Comment by v-shenoy** [comment]
> @v-shenoy would be great if you can take a look on this, more eyes more see. We would like to merge this soon and follow up.

Sorry it was night time here, so I didn't check.

File: pkg/scalers/scaler.go
Comment (v-shenoy) [inline pkg/scalers/scaler.go:46]: This change was missed in the docs - https://github.com/kedacore/keda-docs/blob/main/content/docs/2.9/concepts/external-scalers.md

File: apis/keda/v1alpha1/indentifier.go
Comment (v-shenoy) [inline apis/keda/v1alpha1/indentifier.go:1]: Should be named `identifier.go`

**Comment by v-shenoy** [comment]
Just added two minor comments. Overall, a great job on this :)

File: pkg/scalers/scaler.go
Comment (zroubalik) [reply pkg/scalers/scaler.go:46]: Good catch, thanks!

**Pull request merged**
